### PR TITLE
add cc.d (translated from cc.h)

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by WalterBright** [comment]
Don't let the size intimidate you, it's just a rote translation of cc.h.

File: src/backend/cc.d [hunk 1]
```diff
@@ -0,0 +1,1674 @@
...
+//enum SC;
+struct Thunk;
+struct token_t;
+//struct param_t;
+//struct block;
+//struct Classsym;
+//struct Nspacesym;
+//struct Outbuffer;
+//struct Aliassym;
+struct dt_t;
```
Comment (yebblies) [inline src/backend/cc.d]: Isn't this defined somewhere else?

File: src/backend/cc.d [hunk 2]
```diff
@@ -0,0 +1,1674 @@
...
+    Symbol* CSlinkage;          // table of forward referenced linkage pragmas
+    list_t CSlist_freelist;     // free list for list package
+    symtab_t* CSpsymtab;        // pointer to current Symbol table
+//#if MEMORYHX
+//    void **CSphx;               // pointer to HX data block
+//#endif
+    char* modname;              // module unique identifier
+}
+
+extern Cstate cstate;
```
Comment (yebblies) [inline src/backend/cc.d]: `__gshared`?

File: src/backend/cc.d [hunk 3]
```diff
@@ -0,0 +1,1674 @@
...
+     BCFprivate    = 4,         // base class is private
+
+     BCFvirtual    = 8,         // base class is virtual
+     BCFvfirst     = 0x10,      // virtual base class, and this is the
+                                // first virtual appearance of it
+     BCFnewvtbl    = 0x20,      // new vtbl generated for this base class
+     BCFvirtprim   = 0x40,      // Primary base class of a virtual base class
+     BCFdependent  = 0x80,      // base class is a dependent type
+}
+enum baseclass_flags_t BCFpmask = BCFpublic | BCFprotected | BCFprivate;
```
Comment (yebblies) [inline src/backend/cc.d]: It looks weird to specify the type here

File: src/backend/cc.d [hunk 4]
```diff
@@ -0,0 +1,1674 @@
...
+    GTbyte          = 0x8000,      // variable is sometimes accessed as
+    SFLread         = 0x40000,     // variable is actually read from
+                                   // (to eliminate dead stores)
+    SFLspill        = 0x80000,     // only in register part of the time
+}
+
+struct Symbol
+{
+//#ifdef DEBUG
+    debug ushort      id;
```
Comment (yebblies) [inline src/backend/cc.d]: This makes me very worried.  We shouldn't be changing the struct size and members offsets when debug is active.

File: src/backend/cc.d (hunk 3 above)
Comment (andralex) [inline src/backend/cc.d]: Otherwise it would be inferred as `int`, right?

File: src/backend/cc.d (hunk 4 above)
Comment (andralex) [inline src/backend/cc.d]: It was the case before so it doesn't pertain to reviewing this.

**Comment by andralex** [comment]
Need to add `__gshared` to the global and we're good.

File: src/backend/cc.d (hunk 2 above)
Comment (WalterBright) [inline src/backend/cc.d]: done

File: src/backend/cc.d (hunk 3 above)
Comment (WalterBright) [inline src/backend/cc.d]: yes

**Comment by WalterBright** [comment]
done

**Comment by andralex** [comment]
Auto-merge toggled on

File: src/backend/cc.d (hunk 1 above)
Comment (yebblies) [inline src/backend/cc.d]: What about this?  Do we now have multiple definitions of dt_t?

Comment (WalterBright) [inline src/backend/cc.d]: Yes. Unfortunately, those .h files are all interrelated and depend on each other. I can't unwind those problems until they're all in D. Another problem I discovered is the macros were hiding forward references. Those'll have to be dealt with piece by piece.

The first step is this one - get a rote translation to compile and in the build system.

Comment (yebblies) [inline src/backend/cc.d]: As long as they do get fixed eventually.  Getting the imports right in D should be a lot easier than getting the header mess working in C++.

Comment (WalterBright) [inline src/backend/cc.d]: Sure. But it turns out we might need to adjust D a bit to allow for stub declarations of structs that are defined elsewhere. This practice is common in C/C++, and these difficulties illustrate a speed bump in translating.

**Pull request merged**